File: dp/graphs/max_area.py
from typing import List


def max_area_of_island(grid: List[List[str]]) -> int:
    rows, cols = len(grid), len(grid[0])
    visited = set()

    def count_cells(row: int, col: int):
        if row == rows or row < 0 or col == cols or col < 0 or grid[row][col] == 0 or (row, col) in visited:
            return 0

        visited.add((row, col))
        return (1 + count_cells(row + 1, col) + count_cells(row - 1, col) + count_cells(row, col + 1) + count_cells(row, col - 1))

    area = 0
    for i in range(rows):
        for j in range(cols):
            area = max(area, count_cells(i, j))

    return area


grid = [
    [0, 1, 1, 0, 1],
    [1, 0, 1, 0, 1],
    [0, 1, 1, 0, 1],
    [0, 1, 0, 0, 1]
]

max_islands = max_area_of_island(grid)
print(f"Max Area: {max_islands}")

File: graphs/src/max-area.ts
function maxAreaOfIsland(grid: Array<Array<number>>) {
  let maxArea = 0;
  let [rows, cols] = [grid.length, grid[0].length];
  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      maxArea = Math.max(maxArea, markVisitAndReturnArea(grid, i, j));
    }
  }
  return maxArea;
}

function markVisitAndReturnArea(
  grid: Array<Array<number>>,
  row: number,
  col: number
) {
  const [rows, cols] = [grid.length, grid[0].length];

  if (
    row < 0 ||
    row >= rows ||
    col < 0 ||
    col >= cols ||
    grid[row][col] === 0
  ) {
    return 0;
  }
  let count = 1;
  grid[row][col] = 0;
  count += markVisitAndReturnArea(grid, row - 1, col);
  count += markVisitAndReturnArea(grid, row + 1, col);
  count += markVisitAndReturnArea(grid, row, col - 1);
  count += markVisitAndReturnArea(grid, row, col + 1);
  return count;
}

const grid2 = [
  [0, 1, 1, 0, 1],
  [1, 0, 1, 0, 1],
  [0, 1, 1, 0, 1],
  [0, 1, 0, 0, 1],
];

const result2 = maxAreaOfIsland(grid2);
console.log(result2);
